src/data_structures/src/merkle_tree.cairo

```diff
@@ -2,63 +2,119 @@
 //!
 //! # Example
 //! ```
-//! use alexandria::data_structures::merkle_tree::MerkleTreeTrait;
+//! use alexandria::data_structures::merkle_tree::{MerkleTree, PedersenHasher, PoseidonHasher, MerkleTreeTrait};
 //!
-//! // Create a new merkle tree instance.
-//! let mut merkle_tree = MerkleTreeTrait::new();
-//! let mut proof = array![];
-//! proof.append(element_1);
-//! proof.append(element_2);
+//! // Create a new merkle tree legacy instance, this version uses the pedersen hash method.
+//! let mut merkle_tree: MerkleTree<PedersenHasher> = MerkleTreeTrait::new();
+//! let mut proof = array![element_1, element_2];
+//! // Compute the merkle root.
+//! let root = merkle_tree.compute_root(leaf, proof);
+//!
+//! // Create a new merkle tree instance, this version uses the poseidon hash method.
+//! let mut merkle_tree: MerkleTree<PoseidonHasher> = MerkleTreeTrait::new();
+//! let mut proof = array![element_1, element_2];
 //! // Compute the merkle root.
 //! let root = merkle_tree.compute_root(leaf, proof);
 
 // Core lib imports
-use array::SpanTrait;
-use hash::LegacyHash;
-use traits::Into;
+use array::{ArrayTrait, SpanTrait};
+use pedersen::PedersenTrait;
+use poseidon::PoseidonTrait;
+use hash::HashStateTrait;
+use traits::{Into, Copy, Drop};
+
+/// Hasher trait.
+
+trait HasherTrait<T> {
+    fn new() -> T;
+    fn hash(ref self: T, data1: felt252, data2: felt252) -> felt252;
+}
+
+
+// Hasher representations.
+
+#[derive(Drop, Copy)]
+struct PedersenHasher {}
+
+#[derive(Drop, Copy)]
+struct PoseidonHasher {}
+
+/// Hasher impls.
+
+impl PedersenHasherImpl of HasherTrait<PedersenHasher> {
+    fn new() -> PedersenHasher {
+        PedersenHasher {}
+    }
+    fn hash(ref self: PedersenHasher, data1: felt252, data2: felt252) -> felt252 {
+        let mut state = PedersenTrait::new(data1);
+        state = state.update(data2);
+        state.finalize()
+    }
+}
+
+impl PoseidonHasherImpl of HasherTrait<PoseidonHasher> {
+    fn new() -> PoseidonHasher {
+        PoseidonHasher {}
+    }
+    fn hash(ref self: PoseidonHasher, data1: felt252, data2: felt252) -> felt252 {
+        let mut state = PoseidonTrait::new();
+        state = state.update(data1);
+        state = state.update(data2);
+        state.finalize()
+    }
+}
 
 /// MerkleTree representation.
+
 #[derive(Drop)]
-struct MerkleTree {}
+struct MerkleTree<T> {
+    hasher: T
+}
 
 /// MerkleTree trait.
-trait MerkleTreeTrait {
+trait MerkleTreeTrait<T> {
     /// Create a new merkle tree instance.
-    fn new() -> MerkleTree;
+    fn new() -> MerkleTree<T>;
     /// Compute the merkle root of a given proof.
-    fn compute_root(ref self: MerkleTree, current_node: felt252, proof: Span<felt252>) -> felt252;
+    fn compute_root(
+        ref self: MerkleTree<T>, current_node: felt252, proof: Span<felt252>
+    ) -> felt252;
```

Comment:
What do you mean? Is this because of the generic type? Otherwise this is a the original version: https://github.com/keep-starknet-strange/alexandria/blob/fa3f7e75c3f4074cd0f47cb80c510b5a0916de20/src/data_structures/src/merkle_tree.cairo#L29